tool/tsh/exec_command.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/gravitational/teleport/lib/client"
+
+	"github.com/gravitational/kingpin"
+	"github.com/gravitational/trace"
+)
+
+// execCommand is a command to execute shell command, while setting Teleport-specific environment variables to use.
+// Can be used to run arbitrary programs by setting shell and controlling the arguments.
+type execCommand struct {
+	shell     string
+	arguments []string
+
+	cmd *kingpin.CmdClause
+}
+
+func findShell(shell string) (string, error) {
+	// if shell is explicit, don't try heuristics.
+	if shell != "" {
+		return exec.LookPath(shell)
+	}
+
+	// try to find usable shell.
+	options := []string{os.Getenv("SHELL"), "bash", "sh"}
+
+	for _, option := range options {
+		if option == "" {
+			continue
+		}
+
+		fp, err := exec.LookPath(option)
+		if err == nil {
+			return fp, nil
+		}
+	}
+
+	// nothing worked.
+	return "", trace.BadParameter("unable to find a working shell. set the SHELL env variable or make sure `bash` or `sh` are in PATH.")
```

Comment:
Personally, I find the recommended error capitalization to be a pretty "meh" part of Go. Not great, not terrible - I tend to avoid any capitalized letters in chats and comments... The format of godoc's is a bigger issue IMHO (the first word must be the symbol being documented, which works badly for longer symbols...), but again, not really a deal-breaker. Bit annoying, but not overly so.

To sum it up: I have no strong preferences either way, and I'm happy to change the capitalization if anyone feels strongly about either way.